sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetEncryptionTest.scala

```diff
@@ -0,0 +1,139 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.parquet
+
+import java.io.File
+import java.nio.charset.StandardCharsets
+import java.util.{Base64, HashMap, Map}
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.parquet.crypto.{KeyAccessDeniedException, ParquetCryptoRuntimeException}
+import org.apache.parquet.crypto.keytools.{KeyToolkit, KmsClient}
+import org.apache.spark.sql.QueryTest
+import org.apache.spark.sql.test.SharedSparkSession
+
+import scala.sys.process._
+
+/**
+ * A test suite that tests parquet modular encryption usage in Spark.
+ */
+class ParquetEncryptionTest
+  extends QueryTest with SharedSparkSession {
+
+  private val encoder = Base64.getEncoder
+  private val footerKey =  encoder.encodeToString("0123456789012345".getBytes(StandardCharsets.UTF_8))
+  private val key1 = encoder.encodeToString("1234567890123450".getBytes(StandardCharsets.UTF_8))
+  private val key2 = encoder.encodeToString("1234567890123451".getBytes(StandardCharsets.UTF_8))
+
+  import testImplicits._
+
+  test("Write and read an encrypted parquet") {
+    withTempDir { dir =>
+      spark.conf.set("parquet.crypto.factory.class",
+        "org.apache.parquet.crypto.keytools.PropertiesDrivenCryptoFactory")
+      spark.conf.set("parquet.encryption.kms.client.class",
+        "org.apache.spark.sql.execution.datasources.parquet.InMemoryKMS")
+      spark.conf.set("parquet.encryption.key.list",
+        s"footerKey: ${footerKey}, key1: ${key1}, key2: ${key2}")
+
+      val df = Seq((1, 22, 333)).toDF("a", "b", "c")
+      val parquetDir = new File(dir, "parquet").getCanonicalPath
+      df.write
+        .option("parquet.encryption.column.keys", "key1: a, b; key2: c")
+        .option("parquet.encryption.footer.key", "footerKey")
+        .parquet(parquetDir)
+
+      val parquetPartitionFile = Seq("ls", "-tr", parquetDir).!!.split("\\s+")(0)
+      val fullFilename = parquetDir + "/" + parquetPartitionFile
+      val magic = Seq("tail", "-c", "4", fullFilename).!!
+      assert(magic.stripLineEnd.trim() == "PARE")
+
+      val parquetDF = spark.read.parquet(parquetDir)
+      assert(parquetDF.inputFiles.nonEmpty)
+      val ds = parquetDF.select("a", "b", "c")
+      ds.show()
+    }
+  }
+}
+
+/**
+ * This is a mock class, built just for parquet encryption testing in Spark
+ * and based on InMemoryKMS in parquet-hadoop tests.
+ * Don't use it as an example of a KmsClient implementation.
```

Comment:
It might be helpful to add a link to a sample KmsClient in this comment,
https://github.com/apache/parquet-mr/blob/apache-parquet-1.12.0/parquet-hadoop/src/test/java/org/apache/parquet/crypto/keytools/samples/VaultClient.java